File: web/src/services/ssh-key.ts
import request from '../utils/request'
import type { PageResult } from './audit'

export interface SSHKeyItem {
  id: number
  name: string
  public_key: string
  fingerprint: string
  key_type: string
  comment: string
  created_at: string
}

export interface SSHKeyBinding {
  id: number
  ssh_key_id: number
  asset_id: number
  hostname: string
  ip: string
  username: string
  status: string
  deployed_at: string
}

export interface CreateSSHKeyParams {
  name: string
  public_key?: string
  private_key?: string
  key_type?: string
  comment?: string
}

export interface DeploySSHKeyParams {
  asset_ids: number[]
  username: string
}

export function getSSHKeys(params?: { keyword?: string; page?: number; page_size?: number }): Promise<PageResult<SSHKeyItem>> {
  return request.get('/sshkeys', { params })
}

export async function getAllSSHKeys(): Promise<SSHKeyItem[]> {
  const res: PageResult<SSHKeyItem> = await request.get('/sshkeys', { params: { page: 1, page_size: 200 } })
  return res.list || []
}

export function createSSHKey(data: CreateSSHKeyParams): Promise<SSHKeyItem> {
  return request.post('/sshkeys', data)
}

export function deleteSSHKey(id: number): Promise<void> {
  return request.delete(`/sshkeys/${id}`)
}

export function deploySSHKey(id: number, data: DeploySSHKeyParams): Promise<void> {
  return request.post(`/sshkeys/${id}/deploy`, data)
}

export function getSSHKeyBindings(id: number): Promise<SSHKeyBinding[]> {
  return request.get(`/sshkeys/${id}/bindings`)
}

// 撤销（回收）SSH Key
export interface RevokeSSHKeyParams {
  asset_ids: number[]
  username: string
}

export interface RevokeResultItem {
  asset_id: number
  hostname: string
  ip: string
  status: string
  error?: string
}

export interface RevokeResponse {
  total: number
  success: number
  failed: number
  results: RevokeResultItem[]
}

export function revokeSSHKey(id: number, data: RevokeSSHKeyParams): Promise<RevokeResponse> {
  return request.post(`/sshkeys/${id}/revoke`, data)
}

// SSH Key 轮换
export interface RotateSSHKeyParams {
  new_key_id?: number
  new_name?: string
  new_comment?: string
}

export interface RotateResponse {
  old_key_id: number
  new_key: SSHKeyItem
  deploy: RevokeResponse | null
  revoke: RevokeResponse | null
}

export function rotateSSHKey(id: number, data: RotateSSHKeyParams): Promise<RotateResponse> {
  return request.post(`/sshkeys/${id}/rotate`, data)
}

// 离职清理
export interface DepartureCleanupParams {
  key_ids: number[]
  asset_ids?: number[]
}

export interface DepartureCleanupResponse {
  total_keys: number
  total_assets: number
  revoke_results: RevokeResponse[]
}

export function departureCleanup(data: DepartureCleanupParams): Promise<DepartureCleanupResponse> {
  return request.post('/sshkeys/departure-cleanup', data)
}

// 下载 SSH Key（含解密私钥）
export interface SSHKeyDownloadResponse {
  name: string
  public_key: string
  private_key: string
}

export function downloadSSHKey(id: number): Promise<SSHKeyDownloadResponse> {
  return request.get(`/sshkeys/${id}/download`)
}

// 部署日志
export interface SSHKeyDeployLog {
  id: number
  ssh_key_id: number
  ssh_key_name: string
  asset_id: number
  hostname: string
  ip: string
  username: string
  action: string
  status: string
  error?: string
  operator_id: number
  created_at: string
}

export function getSSHKeyDeployLogs(id: number, params?: { page?: number; page_size?: number }): Promise<PageResult<SSHKeyDeployLog>> {
  return request.get(`/sshkeys/${id}/deploy-logs`, { params })
}

File: internal/dto/ssh_key.go
package dto

// CreateSSHKeyRequest 创建 SSH Key 请求
type CreateSSHKeyRequest struct {
	Name       string `json:"name" binding:"required,max=128"`
	PublicKey  string `json:"public_key"`  // 导入时提供
	PrivateKey string `json:"private_key"` // 导入时提供
	KeyType    string `json:"key_type"`    // rsa / ed25519，生成时使用
	Comment    string `json:"comment"`
}

// DeploySSHKeyRequest 部署 SSH Key 请求
type DeploySSHKeyRequest struct {
	AssetIDs []uint `json:"asset_ids" binding:"required"`
	Username string `json:"username" binding:"required"`
}

// SSHKeyResponse SSH Key 响应
type SSHKeyResponse struct {
	ID          uint   `json:"id"`
	Name        string `json:"name"`
	PublicKey   string `json:"public_key"`
	Fingerprint string `json:"fingerprint"`
	KeyType     string `json:"key_type"`
	Comment     string `json:"comment"`
	CreatedAt   string `json:"created_at"`
}

// SSHKeyBindingResponse SSH Key 绑定响应
type SSHKeyBindingResponse struct {
	ID         uint   `json:"id"`
	SSHKeyID   uint   `json:"ssh_key_id"`
	AssetID    uint   `json:"asset_id"`
	Hostname   string `json:"hostname"`
	IP         string `json:"ip"`
	Username   string `json:"username"`
	Status     string `json:"status"`
	DeployedAt string `json:"deployed_at"`
}

// RevokeSSHKeyRequest 撤销（回收）SSH Key 请求
type RevokeSSHKeyRequest struct {
	AssetIDs []uint `json:"asset_ids" binding:"required"` // 目标资产列表
	Username string `json:"username" binding:"required"`   // 要从哪个用户的 authorized_keys 移除
}

// RotateSSHKeyRequest SSH Key 轮换请求
type RotateSSHKeyRequest struct {
	NewKeyID   *uint  `json:"new_key_id"`  // 使用已有密钥替换，留空则自动生成
	NewName    string `json:"new_name"`    // 新密钥名称（自动生成时使用），留空则自动生成
	NewComment string `json:"new_comment"` // 新密钥备注（自动生成时使用）
}

// DepartureCleanupRequest 离职清理请求
type DepartureCleanupRequest struct {
	KeyIDs   []uint `json:"key_ids" binding:"required,min=1"` // 要清理的 SSH Key ID 列表
	AssetIDs []uint `json:"asset_ids"`                        // 指定资产列表，留空则清理所有已绑定资产
}

// RevokeResultItem 撤销结果项
type RevokeResultItem struct {
	AssetID  uint   `json:"asset_id"`
	Hostname string `json:"hostname"`
	IP       string `json:"ip"`
	Status   string `json:"status"` // success / failed
	Error    string `json:"error,omitempty"`
}

// RevokeResponse 撤销结果响应
type RevokeResponse struct {
	Total   int                `json:"total"`
	Success int                `json:"success"`
	Failed  int                `json:"failed"`
	Results []RevokeResultItem `json:"results"`
}

// RotateResponse SSH Key 轮换响应
type RotateResponse struct {
	OldKeyID uint            `json:"old_key_id"`
	NewKey   *SSHKeyResponse `json:"new_key"`
	Deploy   *RevokeResponse `json:"deploy"` // 新密钥部署结果
	Revoke   *RevokeResponse `json:"revoke"` // 旧密钥撤销结果
}

// DepartureCleanupResponse 离职清理响应
type DepartureCleanupResponse struct {
	TotalKeys     int              `json:"total_keys"`
	TotalAssets   int              `json:"total_assets"`
	RevokeResults []RevokeResponse `json:"revoke_results"`
}

// SSHKeyDownloadResponse SSH Key 下载响应（含解密后的私钥）
type SSHKeyDownloadResponse struct {
	Name       string `json:"name"`
	PublicKey  string `json:"public_key"`
	PrivateKey string `json:"private_key"`
}

// SSHKeyDeployLogResponse 部署日志响应
type SSHKeyDeployLogResponse struct {
	ID         uint   `json:"id"`
	SSHKeyID   uint   `json:"ssh_key_id"`
	SSHKeyName string `json:"ssh_key_name"`
	AssetID    uint   `json:"asset_id"`
	Hostname   string `json:"hostname"`
	IP         string `json:"ip"`
	Username   string `json:"username"`
	Action     string `json:"action"`
	Status     string `json:"status"`
	Error      string `json:"error,omitempty"`
	OperatorID uint   `json:"operator_id"`
	CreatedAt  string `json:"created_at"`
}
